<a href="https://colab.research.google.com/github/fossilofshelfish/text-analysis-final/blob/main/Final_Project_for_text_analysis_in_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduction**
Artificial intelligence, especially deepfake technology, represents a frontier of synthetic media and has profound social implications. While video and image editing used to be a highly demanding skill, AI now allows inexperienced individuals to create indistinguishable deepfake videos simply by providing verbal descriptions. Public discussion surrounding this technology is complex and diverse. Some people are surprised by its creative potential, and some are concerned about its ethical and security risks. YouTube, as the premier platform for video content and community interaction, is an important place where diverse viewpoints converge.

This project employs computational text analysis to investigate the nature of public discussion around "AI deepfake" videos on YouTube. This project aims to answer three central questions: (1) What is the overall sentiment landscape (positive, negative, or neutral) of viewer comments? (2) If a video's popularity affects the tone of its comments? (3) What specific topics people discuss most frequently when talking about deepfakes?

# **Data Collection**
Data collection was accomplished via the YouTube Data API v3. The process began by searching for the keyword "AI deepfake" and sorting the results by relevance, collecting 60 videos as a seed sample. Next, detailed metadata was extracted from these videos, including title, description, channel, publication date, number of views, number of likes, and number of comments. Finally, the top comments for each accessible video were obtained.

# **Methodology**
*   **Sentiment Analysis:** This project used NLTK's VADER sentiment analyzer, which is designed for social media text. It gave each comment, title, and description a compound score from -1 (very negative) to +1 (very positive). Comments were categorized as positive (score > 0.05), negative (score < -0.05), or neutral based on these scores.
*   **Text Mining & Visualization:** To gain an intuitive overview, a word cloud was generated from the aggregated comment text.
*   **Topic Modeling:** To find common themes in the discussions, this project combined video titles with their comments, cleaned the text (removing URLs, special characters, and common stopwords), and applied TF-IDF vectorization followed by Non-Negative Matrix Factorization (NMF) to identify three main topics.
*   **TF-IDF:** latent themes were extracted using a Non-Negative Matrix Factorization (NMF) model, configured for three topics.



In [ ]:
# Keep pandas (already in Colab). Ensure latest NLTK.
!pip -q install --upgrade nltk

import pandas as pd
import nltk
from nltk.sentiment import SentimentIntensityAnalyzer

# Download the VADER lexicon (only needs to run once per runtime)
nltk.download('vader_lexicon')

sia = SentimentIntensityAnalyzer()

import os
from getpass import getpass

# Paste API key when prompted (input will be hidden in Colab)
os.environ["YOUTUBE_API_KEY"] = getpass("Paste your API Key: ")

# Quick sanity check
assert os.environ.get("YOUTUBE_API_KEY"), "API key not set — please run the cell and paste your key."


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 18.5 MB/s eta 0:00:00


[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


Paste your API Key: ··········


In [ ]:

!pip -q install requests tqdm

import os
import json
from urllib.parse import urlencode

import requests
import pandas as pd
from tqdm import tqdm


In [ ]:

API_KEY = os.environ.get("YOUTUBE_API_KEY")
BASE_URL = "https://www.googleapis.com/youtube/v3"

if not API_KEY:
    raise ValueError("Missing API key. Set os.environ['YOUTUBE_API_KEY'] first.")

def yt_get(resource: str, params: dict) -> dict:
    """Call YouTube Data API v3.
    - resource: e.g., 'search', 'videos', 'commentThreads'
    - params: dict of query params (we append the API key here)
    Returns parsed JSON as a Python dict.
    """
    q = {**params, "key": API_KEY}
    url = f"{BASE_URL}/{resource}?{urlencode(q)}"
    r = requests.get(url, timeout=30)
    r.raise_for_status()  # raise an HTTPError if the request failed
    return r.json()


In [ ]:

# Topic on AI deepfake
QUERY = "AI deepfake"
TARGET_VIDEOS = 60           # upper bound of total videos to collect (keep modest: quotas!)
MAX_RESULTS = 50             # per-page limit for search endpoint

video_hits = []              # will hold basic search results
page_token = None            # used for pagination

with tqdm(total=TARGET_VIDEOS, desc="Searching videos") as pbar:
    while len(video_hits) < TARGET_VIDEOS:
        # The 'search' resource finds videos; we request snippet data (title, channel, publishedAt).
        params = {
            "part": "snippet",
            "q": QUERY,
            "type": "video",
            "maxResults": MAX_RESULTS,
            "order": "relevance",
        }
        if page_token:
            params["pageToken"] = page_token

        data = yt_get("search", params)
        items = data.get("items", [])

        for it in items:
            vid = it.get("id", {}).get("videoId")
            if not vid:
                continue
            snip = it.get("snippet", {})
            video_hits.append({
                "video_id": vid,
                "publishedAt": snip.get("publishedAt"),
                "title": snip.get("title"),
                "channelId": snip.get("channelId"),
                "channelTitle": snip.get("channelTitle"),
            })
            pbar.update(1)
            if len(video_hits) >= TARGET_VIDEOS:
                break

        page_token = data.get("nextPageToken")
        if not page_token:
            break  # no more pages

videos_df = pd.DataFrame(video_hits)
videos_df.head(3)


Searching videos: 100%|██████████| 60/60 [00:00<00:00, 74.06it/s]


video_id           publishedAt  \
0  7akzhpx0EIU  2024-09-02T08:59:53Z   
1  uqnIjfe1Ems  2025-06-15T22:29:49Z   
2  Fyzr6jbSI_I  2023-09-01T15:00:10Z   

                                               title  \
0      How to spot deepfakes and AI-generated images   
1  Can You Tell This Isn’t Real? Scammers Hope No...   
2  Deepfake AI – How Dangerous Is It? #shorts #de...   

                  channelId   channelTitle  
0  UCt5OA3LingpZBeEyPYmputQ  New Scientist  
1  UC6uZzgoiYRCBNe3cJp7ZFBA    floydbishop  
2  UCn8zNIfYAQNdrFRrr8oibKw           VICE

In [ ]:

# We'll call 'videos.list' to fetch details for batches of IDs (up to 50 per call)
def chunked(seq, size):
    for i in range(0, len(seq), size):
        yield seq[i:i+size]

video_ids = videos_df["video_id"].dropna().unique().tolist()

video_details = []
for batch in tqdm(list(chunked(video_ids, 50)), desc="Fetching video details"):
    params = {
        "part": "snippet,statistics",
        "id": ",".join(batch),
        "maxResults": 50,
    }
    data = yt_get("videos", params)
    for it in data.get("items", []):
        snip = it.get("snippet", {})
        stats = it.get("statistics", {})
        video_details.append({
            "video_id": it.get("id"),
            "title": snip.get("title"),
            "description": snip.get("description"),
            "publishedAt": snip.get("publishedAt"),
            "channelTitle": snip.get("channelTitle"),
            # Cast numeric strings to integers when possible
            "viewCount": int(stats.get("viewCount", 0) or 0),
            "likeCount": int(stats.get("likeCount", 0) or 0),
            "commentCount": int(stats.get("commentCount", 0) or 0),
        })

video_details_df = pd.DataFrame(video_details)
video_details_df.head(3)


Fetching video details: 100%|██████████| 2/2 [00:00<00:00, 10.88it/s]


video_id                                              title  \
0  7akzhpx0EIU      How to spot deepfakes and AI-generated images   
1  uqnIjfe1Ems  Can You Tell This Isn’t Real? Scammers Hope No...   
2  Fyzr6jbSI_I  Deepfake AI – How Dangerous Is It? #shorts #de...   

                                         description           publishedAt  \
0  It can sometimes be difficult to spot AI-gener...  2024-09-02T08:59:53Z   
1  🚨 Deepfake scams and AI voice cloning are expl...  2025-06-15T22:29:49Z   
2  In a world where an altered photo, video, or v...  2023-09-01T15:00:10Z   

    channelTitle  viewCount  likeCount  commentCount  
0  New Scientist      63024        541            83  
1    floydbishop     145242       2108           403  
2           VICE     362445       6725           160

In [ ]:

# Some videos disable comments. We'll handle HTTP errors gracefully and cap per‑video volume.
all_comments = []

for vid in tqdm(video_details_df["video_id"].tolist(), desc="Fetching comments"):
    page_token = None
    fetched = 0
    try:
        while True:
            params = {
                "part": "snippet",
                "videoId": vid,
                "maxResults": 100,  # API max per page for commentThreads
                "order": "relevance",  # try 'time' if you want chronological
                # 'textFormat': 'plainText' is default
            }
            if page_token:
                params["pageToken"] = page_token

            data = yt_get("commentThreads", params)
            items = data.get("items", [])

            for it in items:
                top = it.get("snippet", {}).get("topLevelComment", {})
                s = top.get("snippet", {})
                all_comments.append({
                    "video_id": vid,
                    "comment_id": top.get("id"),
                    "author": s.get("authorDisplayName"),
                    "publishedAt": s.get("publishedAt"),
                    "likeCount": s.get("likeCount", 0),
                    "text": s.get("textOriginal", ""),
                })
                fetched += 1

            page_token = data.get("nextPageToken")
            if not page_token:
                break  # no more pages

            if fetched >= 300:
                break  # safety cap so a single video doesn’t eat your quota

    except requests.HTTPError as e:
        print(f"Skipping {vid} due to HTTP error: {e}")
        continue

comments_df = pd.DataFrame(all_comments)
comments_df.head(3)


Fetching comments:   7%|▋         | 4/57 [00:01<00:24,  2.19it/s]

Skipping B4jNttRvbpU due to HTTP error: 403 Client Error: Forbidden for url: https://www.googleapis.com/youtube/v3/commentThreads?part=snippet&videoId=B4jNttRvbpU&maxResults=100&order=relevance&key=AIzaSyBf2LFJ221kpMfMbdefPuasT_ohoMHiyD8


Fetching comments: 100%|██████████| 57/57 [00:23<00:00,  2.44it/s]


video_id                  comment_id        author  \
0  7akzhpx0EIU  Ugys4Vu4l2wWP1wvQqt4AaABAg  @Shengpretty   
1  7akzhpx0EIU  UgzmV5TU8AeB8XmVx2l4AaABAg     @nspector   
2  7akzhpx0EIU  Ugz4yOYk9oUn1b5oaG54AaABAg        @Obrik   

            publishedAt  likeCount  \
0  2025-11-19T20:11:15Z          4   
1  2025-06-29T04:39:44Z         23   
2  2025-07-15T04:07:38Z          6   

                                                text  
0  I used Undetectable AI Image Detector earlier ...  
1  This video did not show us how to spot deepfak...  
2  There are a couple tips near the start but not...

In [ ]:

# Already upgraded earlier; safe to re‑run if needed
!pip -q install --upgrade nltk
import nltk
from nltk.sentiment import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')
sia = SentimentIntensityAnalyzer()


[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [ ]:

# Helper to score a text string and return only the 'compound' score ([-1, 1])
def compound_score(text):
    return sia.polarity_scores(text or "")["compound"]

# Video titles & descriptions
video_details_df["title_compound"] = video_details_df["title"].fillna("").apply(compound_score)
video_details_df["description_compound"] = video_details_df["description"].fillna("").apply(compound_score)

# Comments (if any)
if not comments_df.empty:
    comments_df["compound"] = comments_df["text"].fillna("").apply(compound_score)

   # Caculate the average
    overall_sentiment_mean = comments_df["compound"].mean()
    print(f"所有评论的总体情绪平均值 (compound score): {overall_sentiment_mean:.3f}")

    POS, NEG = 0.05, -0.05
    comments_df["sentiment_label"] = comments_df["compound"].apply(
        lambda c: "pos" if c > POS else ("neg" if c < NEG else "neu")
    )

    sentiment_dist = comments_df["sentiment_label"].value_counts(normalize=True) * 100
    print("\n情绪分布百分比:")
    for label in ["pos", "neu", "neg"]:
        if label in sentiment_dist:
            print(f"  {label}: {sentiment_dist[label]:.1f}%")
else:
    print("评论数据为空，无法计算情绪平均值")
    overall_sentiment_mean = None


所有评论的总体情绪平均值 (compound score): 0.028

情绪分布百分比:
  pos: 36.6%
  neu: 31.3%
  neg: 32.1%


In [ ]:
# Common VADER thresholds for sentiment classification
POS, NEG = 0.05, -0.05

if not comments_df.empty:
    # Check if sentiment_label column already exists to avoid duplicate creation
    if "sentiment_label" not in comments_df.columns:
        comments_df["sentiment_label"] = comments_df["compound"].apply(
            lambda c: "pos" if c > POS else ("neg" if c < NEG else "neu")
        )

    # Calculate overall sentiment mean across ALL comments
    overall_sentiment_mean = comments_df["compound"].mean()
    print(f"Overall sentiment mean for all comments: {overall_sentiment_mean:.3f}")

    # Calculate sentiment distribution percentages
    sentiment_dist = comments_df["sentiment_label"].value_counts(normalize=True) * 100
    print("\nSentiment distribution percentages:")
    for label, value in sentiment_dist.items():
        print(f"  {label}: {value:.1f}%")

    # Aggregate sentiment statistics by video
    agg = (comments_df.groupby("video_id").agg(
        n_comments=("comment_id", "count"),
        mean_compound=("compound", "mean"),
        pct_pos=("sentiment_label", lambda s: (s == "pos").mean()),
        pct_neg=("sentiment_label", lambda s: (s == "neg").mean()),
        pct_neu=("sentiment_label", lambda s: (s == "neu").mean()),
    ).reset_index())
else:
    # Create empty placeholder DataFrame for merging if no comments exist
    agg = pd.DataFrame(columns=["video_id", "n_comments", "mean_compound", "pct_pos", "pct_neg", "pct_neu"])
    overall_sentiment_mean = None
    print("No comment data available")

# Merge video details with aggregated comment sentiment data
summary = (
    video_details_df.merge(agg, on="video_id", how="left")
    .assign(
        title_compound=lambda d: d["title_compound"].round(3),
        description_compound=lambda d: d["description_compound"].round(3),
        mean_compound=lambda d: d["mean_compound"].round(3),
        pct_pos=lambda d: (d["pct_pos"]*100).round(1),
        pct_neg=lambda d: (d["pct_neg"]*100).round(1),
        pct_neu=lambda d: (d["pct_neu"]*100).round(1),
    )
)

print(f"\n=== OVERALL STATISTICS ===")
if overall_sentiment_mean is not None:
    print(f"Total comments: {len(comments_df):,}")
    print(f"Total videos analyzed: {len(summary):,}")
    print(f"Overall sentiment mean: {overall_sentiment_mean:.3f}")
    print(f"Sentiment interpretation: {'Positive' if overall_sentiment_mean > 0.05 else 'Slightly positive' if overall_sentiment_mean > 0 else 'Neutral' if overall_sentiment_mean > -0.05 else 'Negative'}")
    print(f"Video sentiment range: {summary['mean_compound'].min():.3f} (most negative) to {summary['mean_compound'].max():.3f} (most positive)")
    print(f"Video sentiment median: {summary['mean_compound'].median():.3f}")

# Define columns to display in the summary
summary_cols = [
    "video_id", "channelTitle", "publishedAt", "viewCount", "likeCount", "commentCount",
    "title_compound", "description_compound", "n_comments", "mean_compound",
    "pct_pos", "pct_neg", "pct_neu", "title"
]

# Display top 10 most positive videos
print(f"\n=== TOP 10 MOST POSITIVE VIDEOS ===")
top10_positive = summary[summary_cols].sort_values(by=["mean_compound"], ascending=False).head(10)

# Display only the essential columns
display_cols = ["title", "mean_compound", "pct_pos", "pct_neg", "pct_neu", "viewCount"]
print(top10_positive[display_cols].to_string(index=False))

# Display bottom 10 most negative videos
print(f"\n=== TOP 10 MOST NEGATIVE VIDEOS ===")
top10_negative = summary[summary_cols].sort_values(by=["mean_compound"], ascending=True).head(10)
print(top10_negative[display_cols].to_string(index=False))

# Calculate and display key insights
if overall_sentiment_mean is not None:
    print(f"\n=== KEY INSIGHTS ===")
    avg_positive_videos = top10_positive["mean_compound"].mean()
    avg_negative_videos = top10_negative["mean_compound"].mean()
    print(f"Average sentiment of top 10 positive videos: {avg_positive_videos:.3f}")
    print(f"Average sentiment of top 10 negative videos: {avg_negative_videos:.3f}")
    print(f"Sentiment gap between best and worst: {top10_positive['mean_compound'].iloc[0]:.3f} - {top10_negative['mean_compound'].iloc[0]:.3f} = {top10_positive['mean_compound'].iloc[0] - top10_negative['mean_compound'].iloc[0]:.3f}")

    # Check correlation between views and sentiment
    if "viewCount" in summary.columns and "mean_compound" in summary.columns:
        correlation = summary[["viewCount", "mean_compound"]].corr().iloc[0, 1]
        print(f"Correlation between views and sentiment: {correlation:.3f}")
        if abs(correlation) < 0.1:
            print("Interpretation: View count has little to no relationship with comment sentiment.")

# Show complete top 5 positive summary (optional)
print(f"\nComplete summary of top 5 most positive videos:")
display(summary[summary_cols].sort_values(by=["mean_compound"], ascending=False).head(5))

Overall sentiment mean for all comments: 0.028

Sentiment distribution percentages:
  pos: 36.6%
  neg: 32.1%
  neu: 31.3%

=== OVERALL STATISTICS ===
Total comments: 6,593
Total videos analyzed: 57
Overall sentiment mean: 0.028
Sentiment interpretation: Slightly positive
Video sentiment range: -0.630 (most negative) to 0.326 (most positive)
Video sentiment median: 0.009

=== TOP 10 MOST POSITIVE VIDEOS ===
                                                                        title  mean_compound  pct_pos  pct_neg  pct_neu  viewCount
                              HONOR Magic 7 Pro Deepfake AI Detection Preview          0.326     50.0      0.0     50.0    1028517
                                       How AI Deepfake Will Destroy The World          0.245     59.3     22.1     18.6      19860
             Deepfake Challenge 2025 🔥 AI Videos That Fool Everyone (Veo 3.1)          0.169     48.4     25.8     25.8       5431
                                How to spot deepfakes and AI-gene

video_id     channelTitle           publishedAt  viewCount  likeCount  \
44  DUJ5BfCAEfk           1side0  2024-12-16T09:27:59Z    1028517        253   
46  AEyFu1OJvuM       Louis Zhao  2025-06-22T21:25:10Z      19860       1376   
15  kNEPV_qYZ1U      floydbishop  2025-10-25T12:01:24Z       5431        180   
0   7akzhpx0EIU    New Scientist  2024-09-02T08:59:53Z      63024        541   
51  wQfeYdjcfN4  Niccolò Zanella  2024-12-12T21:18:49Z       8397        134   

    commentCount  title_compound  description_compound  n_comments  \
44             6           0.604                 0.774         6.0   
46           196          -0.542                 0.748       140.0   
15            73          -0.382                -0.986        31.0   
0             83           0.000                 0.868        59.0   
51             7           0.210                 0.976         6.0   

    mean_compound  pct_pos  pct_neg  pct_neu  \
44          0.326     50.0      0.0     50.0   
46          0.245     59.3     22.1     18.6   
15          0.169     48.4     25.8     25.8   
0           0.161     47.5     22.0     30.5   
51          0.146     50.0     33.3     16.7   

                                                title  
44    HONOR Magic 7 Pro Deepfake AI Detection Preview  
46             How AI Deepfake Will Destroy The World  
15  Deepfake Challenge 2025 🔥 AI Videos That Fool ...  
0       How to spot deepfakes and AI-generated images  
51  HONOR Magic7 Pro Ai Deep Fake recognition #HON...

# Sentiment Analysis
The sentiment results show clear patterns in how people react to different types of deepfake content. Videos that teach detection methods or explain how to spot fakes receive the most positive comments (like the HONOR Magic 7 Pro preview with 0.326 sentiment). People appreciate practical knowledge that helps them protect themselves. On the other hand, videos that focus on dangers or misuse scenarios—especially those about job interviews or political misinformation—get very negative reactions (-0.630 sentiment). The nearly equal split between positive, negative and neutral comments (about one-third each) confirms that people have mixed feelings about this technology, not strong agreement in one direction.


In [ ]:
# Scatter: Relationship between viewCount and mean comment sentiment
if 'summary' in globals() and not summary.empty and summary['mean_compound'].notna().any():
    scatter_df = summary.dropna(subset=["mean_compound"]).copy()
    # Use log scale for views if counts vary widely
    fig_scatter = px.scatter(
        scatter_df,
        x="viewCount",
        y="mean_compound",
        hover_name="title",
        hover_data=["channelTitle", "likeCount", "n_comments"],
        title="View count vs. mean comment sentiment",
        labels={"viewCount": "Views", "mean_compound": "Mean compound sentiment"},
    )
    fig_scatter.update_xaxes(type="log")

    fig_scatter.show()

    # Save HTML + PNG
    fig_scatter.write_html("plot_views_vs_sentiment.html", include_plotlyjs="cdn", full_html=True)
    fig_scatter.write_image("plot_views_vs_sentiment.png")
else:
    print("No sentiment summary to plot. Ensure the aggregation step ran successfully.")


# Relationship between viewcount and mean comment sentiment
When I plotted view count against average comment sentiment, no clear pattern emerged. Videos with millions of views had comment sentiment similar to videos with only thousands of views. However, the most extremely positive comment sections weren't typically found on the most-viewed videos. This indicates that smaller communities around niche content might foster more consistently positive discussions, while viral videos attract more diverse opinions that average out to more neutral sentiment.

In [ ]:

# Export tidy CSVs for later analysis or visualization
videos_df.to_csv("videos_search_hits.csv", index=False)
video_details_df.to_csv("video_details.csv", index=False)
comments_df.to_csv("video_comments.csv", index=False)
summary.to_csv("video_sentiment_summary.csv", index=False)

print("Saved: videos_search_hits.csv, video_details.csv, video_comments.csv, video_sentiment_summary.csv")


In [ ]:
# --- WORD CLOUD (comments text) ---
#!pip -q install wordcloud
import nltk
nltk.download('stopwords')
from wordcloud import WordCloud
import matplotlib.pyplot as plt
from nltk.corpus import stopwords

# Combine all comments into one long string
text_all = " ".join(comments_df["text"].dropna().tolist())

# stopwords
EN_STOP = set(stopwords.words('english'))
CUSTOM_STOP = {
    "https", "http", "www", "com", "video", "like", 'one', "can't",
    "really", "could", "also", "even", "would", "thing", "much", "anything",
    "say", "year", "get", "got", "someone", "said", "always", "way", "still",
    "u", "first", "look", "see", "every", "years", "know", "actually", "word",
    "make", "guy", "give", "things", "call", "never", "time", "god", "something"
}
STOPWORDS = EN_STOP.union(CUSTOM_STOP)

# Create word cloud
wc = WordCloud(
    width=1200,
    height=800,
    background_color="white",
    stopwords=STOPWORDS,
    collocations=False
).generate(text_all)

plt.figure(figsize=(14, 9))
plt.imshow(wc, interpolation="bilinear")
plt.axis("off")
plt.title("Word Cloud of All YouTube Comments on Deepfake Videos")
plt.show()


# Word Cloud Analysis
The word cloud reveals three dominant strands of public discourse surrounding deepfake technology.

**Authenticity and Trust:** The prominence of terms like "fake," "real," and "believe" highlights a core public anxiety: the erosion of trust in digital media and the challenge of verifying authenticity.

**Misuse and Harm:** Clusters around "use," "people," "bad," "scam," and "problem" underscore a prevailing concern about malicious application. This aligns with the negative reception of tutorial videos, reflecting fear that the technology will be weaponized for fraud and deception.

**Calls for Action:** Words like "need," "talk," and "stop" indicate a demand for intervention. The discourse extends beyond observation to an expectation for societal, regulatory, or technological responses to mitigate risks.

In [ ]:
# --- Topic Modeling on Title + Comments ---

import re
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import NMF
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

# Download 'wordnet' for lemmatization
nltk.download('wordnet')

# stopwords
EN_STOP = set(stopwords.words('english'))
CUSTOM_STOP = {"https", "http", "www", "com", "video", "like", 'one', "can't",
    "really", "could", "also", "even", "would", "thing", "much", "anything",
    "say", "year", "get", "got", "someone", "said", "always", "way", "still",
    "u", "first", "look", "see", "every", "years", "know", "actually", "word",
    "make", "guy", "give", "lol",
    "de", "il", "le", "la", "les", "un", "une", "des", "et", "est",
    "que", "dans", "pour", "par", "sur", "avec", "son", "ses",
    "emmanuelmacron", "bellaciao", "paris", "parodie", "svp", "humour",
    "petite", "macron", "ce",  "qui", "je", "bien", "du", "en", "est", "qu", "pa", "lui", "pi",}
STOPWORDS = EN_STOP.union(CUSTOM_STOP)

lemmatizer = WordNetLemmatizer()

def clean_text(t):
    if pd.isna(t):
        return ""
    t = t.lower()
    t = re.sub(r"http\S+", "", t)
    t = re.sub(r"@\w+", "", t)
    t = re.sub(r"[^a-zA-Z\s']", " ", t)
    tokens = [lemmatizer.lemmatize(w) for w in t.split() if w not in STOPWORDS]
    return " ".join(tokens)

# Merge title from video_details_df into comments_df
# This is necessary because comments_df itself does not have a 'title' column
# Check if 'title' column already exists in comments_df and drop it to avoid merge conflicts on re-run
if 'title' in comments_df.columns:
    comments_df = comments_df.drop(columns=['title'])

comments_df = comments_df.merge(video_details_df[['video_id', 'title']], on='video_id', how='left')

# merge title + comment
comments_df["extended_text"] = comments_df["title"] + " " + comments_df["text"]
comments_df["clean"] = comments_df["extended_text"].apply(clean_text)

# TF-IDF
vectorizer = TfidfVectorizer(max_df=0.9, min_df=20, max_features=8000)
X = vectorizer.fit_transform(comments_df["clean"])
feature_names = vectorizer.get_feature_names_out()

# NMF
n_topics = 3
nmf = NMF(n_components=n_topics, random_state=42)
W = nmf.fit_transform(X)
H = nmf.components_

# print topics
for i, topic in enumerate(H):
    print(f"\n--- TOPIC {i+1} ---")
    top_words = topic.argsort()[-15:]
    print(", ".join([feature_names[j] for j in top_words]))


# TF-IDF and Topic Modeling
The topic modeling revealed three coherent themes in the conversations.

**Topic 1: Spotting Fakes and Avoiding Scams**
Keywords: tell, deepfakes, old, fake, scammed, fool, scammer, real, video
This topic centers on practical concerns about being deceived. People share experiences of being "scammed" and discuss methods to "tell" real from "fake" content. The inclusion of "old" suggests concern about existing videos being manipulated, not just newly created deepfakes.

**Topic 2: Technological Impact and Entertainment**
Keywords: anymore, laugh, scary, movie, realistic, cruise, tom
Here, people express mixed feelings about the technology itself. Words like "anymore" contrasted with "ago" show awareness of rapid technological progress. References to "Tom Cruise" in "movies" highlight entertainment applications, while "laugh" and "scary" capture the dual response of both amusement and concern about increasingly "realistic" outputs.

**Topic 3: Real-World Consequences and Legal Issues**
Keywords: young, lawyer, sue, frame, dhs, black
This topic appears to focus on actual cases and societal impacts. "DHS" refers to the U.S. Department of Homeland Security, indicating discussions about national security and institutional responses. Terms like "lawyer," "sue," and "frame" suggest conversations about legal recourse for victims, while demographic references like "young" and "black" hint at concerns about which groups might be disproportionately affected.


# **Discussion**
This project successfully used text analysis methods to map public opinion on a complex topic. The combination of sentiment analysis, word clouds, and topic modeling provided complementary views of the discussion. Sentiment scores showed emotional reactions, the word cloud revealed frequently discussed concerns (like "fake," "scam," "real"), and topic modeling organized conversations into practical themes. While the dataset had limitations—only 60 videos from one platform—the methods worked well together to show how YouTube commenters are actively debating AI deepfakes rather than simply accepting or rejecting the technology.

# **Limitations**
This project has several limitations worth noting. The dataset of 60 videos, while sufficient for analysis, represents only a small slice of YouTube content on this topic. The VADER sentiment analyzer, though well-suited for social media, may misinterpret sarcasm or nuanced language. The topic modeling approach required subjective decisions, such as choosing three topics, though this number yielded interpretable results. Most significantly, the analysis focused exclusively on viewer comments without examining video content itself. This means we analyzed reactions without fully understanding what prompted them, which might explain why some negatively-framed videos still received positive comments—viewers might have been appreciating important warnings rather than endorsing the technology.

# **Conclusion**
Overall, this analysis shows that public discussion of AI deepfakes focuses on practical questions rather than abstract debates. People are most concerned about how to identify synthetic media, protect themselves from scams, and understand real-world consequences. The positive response to educational content suggests that providing clear information can help audiences navigate this technology more confidently. As deepfake tools become more accessible, this kind of public discourse analysis can help creators, educators and policymakers understand what information people need and what concerns should be addressed.

# Reference
Xu, Z., Wen, X., Zhong, G., & Fang, Q. (2025). Public perception towards deepfake through topic modelling and sentiment analysis of social media data. Social Network Analysis and Mining, 15(16). https://doi.org/10.1007/s13278-025-01445-8